---
title: "Introduction to ngc"
author: "KZ"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to ngc}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
options(scipen = 1, digits = 2) #set to two decimal 
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
```

## Install the package

```{r, eval=TRUE}
# on CRAN
source("defn.R")
source("simulate.r")
source("array2mat.r")
source("pldag.set.r")
source("ngc.r")
source("grangerLasso.r")
source("grangerTLasso_v3.r")
source("grangerThrLasso.r")
source("metrics.R")
library(glmnet)
library(gglasso)
library(igraph)
library(Matrix)
```
# Introduction

`ngc` provides several methods to estimate causal relationships from time-course data and temporal panel data. It mainly implements truncating lasso, threshold lasso method, and group lasso regression regularization framework. Details of these three methods are available in the references. 

In this vignette, we demonstrate the `ngc` package using a simulated dataset. `ngc` can handle two different types of the data, either data with $T$ time points and $p$ concurent time series or data with $n$ observations, $T$ time points and $p$ concurent time series. Also, the dataset is the only object that is required for using `ngc`. This package requires the packages attached before using it:

*   glmnet
*   gglasso
*   igraph
*   Matrix

## Data set-up
To begin with, we demonstrate how to simulate a set of data from the network using function `defn_net` and `simulate_data`. We first creates the skeleton of a random sparse $d \times p \times p$ network with $0.1$ sparsity and $n$ samples, where $d$ is the number of lags considered to generate time series. The value of `sparsity` can be customized, and its default is ranging from [0.01, 0.05]. The `grp` argument takes a vector (numeric or character) and must be the same length as $p$. Each element of the group tells us which group the network $d \times p$ corresponds to. 
The argument `grp_sparsity` controls the sparsity of the groups when users want to generate network with group effects.

In this cases, we consider, $10$ iid samples from the Granger causality network with $16$ variables and $2$ lags, that is $n = 10$, $p = 16$, and $d=2$.

```{r, eval=TRUE, cache=TRUE}
## Creates the skeleton of a random sparse d x p x p network 
## with 0.05 sparsity
set.seed(123)
d <- 2
p <- 16
n <- 10
edge <- defn_net(d = d, p = p, n = n, sparsity = 0.1)
dim(edge)
edge[1,,]
```

Then, we can simulate $n$ iid samples from the Granger causality network we constructed before over $T$ time points. $X$ is a $n \times p \times T$ array. In this example, we set $T= 100$ and the standard deviation of the error term is $0.2$. 

```{r, eval=TRUE, cache=TRUE}
T <- 100
error_sd <- 0.2
X <- simulate_data(n, edge, T = T, error_sd = error_sd)
str(X)
dim(X)
```

Users can either load their own data or use our functions to generate data. Note that the time series simulated from function `simulate_data()' is stationary.

## Methods Implementation

### Lasso
Now that we have our data set-up properly we can begin with the models. `ngc` has three main models. The default model used in the package is the lasso, which set `method = 'regular'`. We will demonstrate in this section. 

#### Regular Lasso
We fit the model using the most basic call to `ngc`, which estimates graphical Granger causality without group effect. We set the lag $d = 2$.
```{r, eval=TRUE, cache=TRUE, warning = FALSE}
fit1 = ngc(X, d=2)
```

A summary of the object `fit1` is displayed if we just enter the object name or use the `print` function. It contains lots of information. To save some space, we don't display here. Instead, we use the `name` function to show the possible information returned by the object `fit1`.
```{r, eval=TRUE, cache=TRUE}
names(fit1)
```

`estMat` gives the estimation of the Granger Causality; `lambda` gives the tuning parameter $\lambda$ for each concurrent time series we considered in the first step; `sigma` returns the threshold parameter $\sigma$ we use for the second step of the method; `intercepts` returns the intercept in the estimation procedure; `tsOrder` returns the order of lags selected in the model; `dag` and `ring` gives the information for estimated the network, and we will use this information for visualization later.
```{r, eval=TRUE, cache=TRUE}
fit1$estMat
fit1$lambda
fit1$sigma
fit1$intercepts
```
Note that user can also specify the value of type I error using the argument `typeIerr`. If type I error is not provided, the tuning parameter is selected by cross-validation, otherwise error-based lasso is fitted.
```{r, eval=TRUE, cache=TRUE, warning = FALSE}
fit1_typeI = ngc(X, d=2, typeIerr = 0.02)
fit1_typeI$estMat
fit1_typeI$lambda
```

`plot.ngc` provides several options for visualizing the results from `ngc`. To visualize the model, we can simply call the function `plot` or `plot.ngc`. `ngc.type` can be used to select the type of graph to show the Granger causality. The default type is `granger`, which plots the DAG of network. Users can also show the ring graph by specifying `ngc.type = "granger"`.
```{r, eval=TRUE, cache=TRUE}
plot.ngc(fit1, ngc.type = "dag")
plot.ngc(fit1, ngc.type = "granger")
```


Users can also make predictions at a given time point. For example, if user set $tp = 2$, the output is the predicted network structure at time $T+2$.
```{r, eval=TRUE, cache=TRUE}
fit1_pred <- predict(fit1, 2)
dim(fit1_pred)
head(fit1_pred)
```

#### Regular Lasso with Grouping
Next, we estimates graphical Granger causality with group effect. Users could specify the group structure using argument `group`. The argument `group` should be consecutive integer with length equal to the number of concurrent time series $p$. For example, if we know that the first half of the time series and the later half of the time series can be separately groupped together, then we can set the group structure as $1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2$.
```{r, eval=TRUE, cache=TRUE}
group = rep(1:2, each = 8)
fit1 = ngc(X, d=2, group = group)
```
Users can also specify the type of grouping using `groupByTime`. The default is FALSE, which means model will not invoke group by time and its lags. If `groupByTime = TRUE`, then Granger causality is estimated using grouping by time.
```{r, eval=TRUE, cache=TRUE}
fit1 = ngc(X, d=2, groupByTime = TRUE)
plot.ngc(fit1, ngc.type = "dag")
```

#### Regular Lasso with Unkown Time Order
If time order is not specified, the model can automatically choose the time order and estimate the Granger Causality correspondingly. 
```{r, eval=TRUE, cache=TRUE, warning=FALSE}
fit1 = ngc(X)
fit1$tsOrder
```


### Threshold Lasso
Now we demonstrate the usage of threshold lasso model. Users can set `method = 'threshold'`. Use the simulated data above. We fit the model to estimates graphical Granger causality without group effect. Here, we choose lag to be 2. User can also specify the value of type I error. If type I error is provided, then error-based lasso is fitted.
```{r, eval=TRUE, cache=TRUE}
fit2 <- ngc(X, d = d, method = 'threshold', typeIerr = 0.02)
```
Similarly to the original Lasso method, we can get the estimation, visualize the model and make predictions. We will first check the estimations.
```{r, eval=TRUE, cache=TRUE}
fit2$estMat
fit2$lambda
fit2$sigma
fit2$intercepts
fit2$tsOrder
```
The estimated order is 0, which means there are no Granger causality in the network. So the visualization of the results using threshold Lasso method have no edges.
```{r, eval=TRUE, cache=TRUE}
plot.ngc(fit2, ngc.type = "dag")
```

Note that from the picture, the threshold Lasso method gives more sparse network.

The predictions can be made at a given time point as well. For example, if user set $tp = 5$, the output is the predicted network structure at time $T+5$.
```{r, eval=TRUE, cache=TRUE}
fit2_pred <- predict(fit2, 5)
dim(fit2_pred)
head(fit2_pred)
```

#### Threshold Lasso with Grouping
Next, we estimates graphical Granger causality with group effect. Similarly to the regular lasso approach, we can set the group structure as $1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2$ and choose group by time lags.
```{r, eval=TRUE, cache=TRUE}
group = rep(1:2, each = 8)
fit2 = ngc(X, d=2, method = 'threshold', typeIerr = 0.02, group = group, groupByTime = TRUE)
plot.ngc(fit2, ngc.type = "dag")
```


#### Threshold Lasso with Unkown Time Order
If time order is not specified, the model can automatically choose the time order and estimate the Granger Causality correspondingly. 
```{r, eval=TRUE, cache=TRUE}
fit2 = ngc(X, method = 'threshold', typeIerr = 0.02)
fit2$tsOrder
```


### Truncated Lasso
Next, we demonstrate the usage of truncated lasso model. Users can set `method = 'truncate'`. Again, apply the simulated data above. We fit the model to estimates graphical Granger causality without group effect. 

It automatically identify the time series order. So users do not need to specify the number of lags. But users can also customize the number of lags as well. The value of type I error is used to select the proper tuning parameter $\lambda$, while the type II error is used to calculate truncating factor $\Phi_t$. If type I error is provided, then error-based lasso is fitted.

Note that this method does not incorporate grouping, so users the arguments `group` and `groupByTime` should set as default,  that is `None` and `FALSE` correspondingly.

Here we first demonstrate truncated lasso method without specifying the order of time series.

```{r, eval=TRUE, cache=TRUE}
fit3 <- ngc(X,  method = 'truncate',
            typeIerr = 0.05)
```
Similarly to the previous methods, we can get the estimations. `fit3$estMat` gives the estimated Granger Causality. `fit3$tsOrder` gives the estimated time series order detected by the method.
```{r, eval=TRUE, cache=TRUE}
fit3$lambda
fit3$intercepts
fit3$tsOrder
```
The visualization and prediction is listed below:
```{r, eval=TRUE, cache=TRUE}
plot.ngc(fit3, ngc.type = "dag")
plot.ngc(fit3, ngc.type = "granger")
```

We show the predicted network structure at time $T+2$.
```{r, eval=TRUE, cache=TRUE}
fit3_pred <- predict(fit3, 2)
dim(fit3_pred)
head(fit3_pred)
```

The user can also customize the order of lags. For example, we  set $d = 3$.
```{r, eval=TRUE, cache=TRUE}
fit4 <- ngc(X, d = 3, method = 'truncate', 
             typeIerr = 0.05)
```
The visualization is shown below:
```{r, eval=TRUE, cache=TRUE}
plot.ngc(fit4, ngc.type = "dag")
plot.ngc(fit4, ngc.type = "granger")
```

In addition, we show the predicted network structure at time $T+2$ as well. The results are the similar.
```{r, eval=TRUE, cache=TRUE}
fit4_pred <- predict(fit4, 2)
dim(fit4_pred)
head(fit4_pred)
```

## Real Data Example
In this section, we will demonstrate our method with real data application. The data is a subset of genome-wide expression of cell cycle genes in human cancer
cell lines (HeLa) downloaded from 





**References**

Shojaie A. and Michailidis G. (2010) Discovering Graphical Granger Causality Using a Truncating Lasso Penalty, Bioinformatics, 26(18): i517-i523

Shojaie A., Basu S. and Michailidis G. (2012) Adaptive Thresholding for Reconstructing Regulatory Networks from Time Course Gene Expression Data, Statistics In Biosciences 4(1): 66-83

Basu S., Shojaie A. and Michailidis G. (2015) Network Granger Causality with Inherent Grouping Structure, Journal of Machine Learning Research (JMLR)
